### ISSUE TITLE
Socks 4/5 Proxy handler in NetClient

### ISSUE THREAD

**Issue by FainTMako** [vert-x3/issues#101]
Hello,

I'm requesting that we are able to use Socks 4/5 proxies on the vertx NetClient.
I propose that we add additional fields to the NetClientOptions class to allow for the easy addition of the handler to the pipeline if needed.

Thank you for your time.

**Comment by cescoffier** [comment]
As a reference, here is a stack overflow question about it:

http://stackoverflow.com/questions/35119032/how-to-use-socks4-5-proxy-handlers-in-netty-client-4-1

```
p.addFirst(new Socks4ProxyHandler(new InetSocketAddress("149.202.68.167", 37678)));
```

**Comment by pmlopes** [comment]
Duplicate of #1361

**Comment by alexlehm** [comment]
not really duplicate but certainly related

**Comment by alexlehm** [comment]
we have finished the implementation of https connect proxy (https://github.com/eclipse/vert.x/issues/1361), I will start the socks proxy implementation

**Comment by alexlehm** [comment]
I have pushed my WIP version, this is not quite correct yet and it requires some better tests plus a bit of clean-up with the codestyle

**Comment by alexlehm** [comment]
I have fixed a bug that caused an issue when connecting with NetClient, I think the operation of the code is now correct, it requires some better tests plus a bit of clean-up with the codestyle and there are way too many log statements left

https://github.com/alexlehm/vert.x/tree/issues/%23101-socks45-support

**Comment by alexlehm** [comment]
@FainTMako do you consider support for Socks4 (not a) necessary? I assume that most servers will use Socks5 and support fallback to Socks4a based on the version number, but Socks4 would require different DNS resolution, which I didn't implement currently

**Comment by alexlehm** [comment]
I squashed the commits into a single commit since there were way too many back and forth changes, its all working now and I have added a bit of docu

https://github.com/alexlehm/vert.x/tree/issues/%23101-socks45-squashed

**Comment by alexlehm** [comment]
added example to the vertx-examples project https://github.com/vert-x3/vertx-examples/pull/130

**Comment by alexlehm** [comment]
Julien restructured the ChannelProvider and the netty handlers a bit, I changed a few small things in comments mostly, the version is on this branch now:
https://github.com/alexlehm/vert.x/tree/socks-proxy-julien

**Comment by alexlehm** [comment]
fixed by pr #1439
